libs/vkd3d/meta.c

```diff
@@ -873,6 +873,274 @@ const struct vkd3d_format *vkd3d_meta_get_copy_image_attachment_format(struct vk
+        ds_state.back = ds_state.front;
+
+        vk_module = meta_resolve_image_ops->vk_fs_stencil_module;
+
+        if (!meta_ops->device->vk_info.EXT_shader_stencil_export)
+            dynamic_state_count = ARRAY_SIZE(dynamic_states);
+    }
+    else if (key->dst_aspect == VK_IMAGE_ASPECT_DEPTH_BIT)
+    {
+        vk_module = meta_resolve_image_ops->vk_fs_depth_module;
+    }
+    else
+    {
+        switch (key->format->type)
+        {
+            case VKD3D_FORMAT_TYPE_TYPELESS:
+                ERR("Typeless format not supported as a resolve target.\n");
+                return E_INVALIDARG;
+
+            case VKD3D_FORMAT_TYPE_SINT:
+                vk_module = meta_resolve_image_ops->vk_fs_sint_module;
+                break;
+
+            case VKD3D_FORMAT_TYPE_UINT:
+                vk_module = meta_resolve_image_ops->vk_fs_uint_module;
+                break;
+
+            default:
+                vk_module = meta_resolve_image_ops->vk_fs_float_module;
+                break;
+        }
+    }
+
+    if ((vr = vkd3d_meta_create_graphics_pipeline(meta_ops,
+            meta_resolve_image_ops->vk_pipeline_layout,
+            has_depth_target ? VK_FORMAT_UNDEFINED : key->format->vk_format,
+            has_depth_target ? key->format->vk_format : VK_FORMAT_UNDEFINED,
+            key->format->vk_aspect_mask,
+            VK_NULL_HANDLE, vk_module, VK_SAMPLE_COUNT_1_BIT,
+            has_depth_target ? &ds_state : NULL,
+            dynamic_state_count, dynamic_states,
+            &spec_info, true, &pipeline->vk_pipeline)) < 0)
+        return hresult_from_vk_result(vr);
+
+    pipeline->key = *key;
+    return S_OK;
+}
+
+HRESULT vkd3d_meta_get_resolve_image_pipeline(struct vkd3d_meta_ops *meta_ops,
+        const struct vkd3d_resolve_image_pipeline_key *key, struct vkd3d_resolve_image_info *info)
+{
+    struct vkd3d_resolve_image_ops *meta_resolve_image_ops = &meta_ops->resolve_image;
+    struct vkd3d_resolve_image_pipeline *pipeline;
+    HRESULT hr;
+    size_t i;
+    int rc;
+
+    if ((rc = pthread_mutex_lock(&meta_resolve_image_ops->mutex)))
+    {
+        ERR("Failed to lock mutex, error %d.\n", rc);
+        return hresult_from_errno(rc);
+    }
+
+    info->vk_set_layout = meta_resolve_image_ops->vk_set_layout;
+    info->vk_pipeline_layout = meta_resolve_image_ops->vk_pipeline_layout;
+    info->needs_stencil_mask = key->dst_aspect == VK_IMAGE_ASPECT_STENCIL_BIT &&
+            !meta_ops->device->vk_info.EXT_shader_stencil_export;
+
+    for (i = 0; i < meta_resolve_image_ops->pipeline_count; i++)
+    {
+        pipeline = &meta_resolve_image_ops->pipelines[i];
+
+        if (!memcmp(key, &pipeline->key, sizeof(*key)))
+        {
+            info->vk_pipeline = pipeline->vk_pipeline;
+            pthread_mutex_unlock(&meta_resolve_image_ops->mutex);
+            return S_OK;
+        }
+    }
+
+    if (!vkd3d_array_reserve((void **)&meta_resolve_image_ops->pipelines, &meta_resolve_image_ops->pipelines_size,
```

Comment:
Does this have to be dynamically? We statically know the upper bound.

Reply:
Do we though? It's something like 2 pipelines for each integer/stencil format and 3 for all other formats, and even ignoring that we still don't know the total format count here.

FWIW, this whole boilerplate is largely identical to the copy image stuff, and that also uses a dynamic array.

Reply:
Ah, right, we key on format, nvm.